annotation/src/main/java/io/smallrye/common/annotation/Blocking.java

```diff
@@ -0,0 +1,24 @@
+package io.smallrye.common.annotation;
+
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.TYPE;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation used to indicate that the annotated element (method or class) is inherently blocking and so should not be
```

Comment:
What's the meaning when a _class_ is annotated? As if all relevant methods were annotated?

Reply:
And does this class annotation apply to subclasses? (IMHO, annotating classes as a shortcut to annotating all relevant methods opens a nasty can of worms. See the Jakarta Interceptors specification for an example :-) )

Reply:
> And does this class annotation apply to subclasses?

Well, it's not annotated with `@Inherited` so it should not apply to subclasses if you follow the general common annotations rules ;-)

Reply:
Maybe we should make it `@Target({ METHOD })` only?

Reply:
> Well, it's not annotated with `@Inherited`

Right, I missed that, sorry about the noise :-)

Reply:
Anyway -- my main point is that the javadoc should answer these questions (or at least those that make sense :-) ).

Reply:
I don't believe we should support inherited classes. That's just too tricky. I will add a line about that. 

About `TYPE`, it's (again) driven by the gRPC use case, where you can only intercept the whole service and not a single method. The initial model we had in mind was:

```java
@Singleton 
@Blocking 
public class MyService extends GrpcGreeterBaseImpl {

}
```

And this was resulting in registering the _blocking server call_ interceptor for the whole service (so all methods). 

A method based `@Blocking` would need a smarter interceptor that only wraps the call to blocking methods. But in the interceptor class, I don't know if the method is blocking as I don't have access to it. UPDATE: I may have found a way...

Reply:
If it's a CDI interceptor, then the `InvocationContext` gives you the intercepted method :-) But I can see how the "command" pattern might require annotating a class (which implements the command).

Reply:
Nope, it's a gRPC Server Interceptor.

Reply:
Ah OK. This essentially means that the annotation meaning is heavily context-dependent. I should have figured that out sooner, sorry, it's Monday :-)

For example, in case of SmallRye Fault Tolerance, where I'd like to augment MicroProfile Fault Tolerance's `@Asynchronous` (which should be called `@ThreadOffload` IMHO :-) ) with `@Blocking` or `@Nonblocking` [1], I can say that the CDI interceptor treats class-level annotations like a shortcut for annotating all relevant methods.

And for gRPC, it would be something else probably.

[1] I can see 2 possible meanings: 1. `@Nonblocking` means no thread offload is really necessary and `@Blocking` is superfluous, 2. `@Nonblocking` means offload to the event loop and `@Blocking` means offload to a worker thread pool.

Reply:
Yes, it is context-dependent, and it will be up to the "integrator" to clearly explain what it does in the scope of what the general documentation (the JavaDoc written here) says. I should actually mention that. 

Note that now you mentioned it, I wonder if I could use a CDI interceptor instead of a gRPC server interceptor. That would simplify A LOT my "try to discover which method/service is called."

Reply:
So, it cannot be a CDI interceptor (because a method can consume and return streams which need specific handling), but not a problem... I've found a way.

* the annotations don't need to a qualifiers (woohoo, I made @mkouba happy for once!)
* the annotations should be `METHOD` only.

Reply:
Yay! :-)